netmaster/mastercfg/policyState.go

```diff
@@ -0,0 +1,388 @@
+	gp.StateDriver = stateStore
+
+	log.Infof("Creating new epg policy: %s", epgpKey)
+
+	// init the dbs
+	gp.RuleMaps = make(map[string]*RuleMap)
+
+	// Install all rules within the policy
+	for ruleKey := range policy.LinkSets.Rules {
+		// find the rule
+		rule := contivModel.FindRule(ruleKey)
+		if rule == nil {
+			log.Errorf("Error finding the rule %s", ruleKey)
+			return nil, core.Errorf("rule not found")
+		}
+
+		log.Infof("Adding Rule %s to epgp policy %s", ruleKey, epgpKey)
+
+		// Add the rule to epg Policy
+		err := gp.AddRule(rule)
+		if err != nil {
+			log.Errorf("Error adding rule %s to epg polict %s. Err: %v", ruleKey, epgpKey, err)
+			return nil, err
+		}
+	}
+
+	// Save the policy state
+	err := gp.Write()
+	if err != nil {
+		return nil, err
+	}
+
+	// Save it in local cache
+	epgPolicyDb[epgpKey] = gp
+
+	log.Info("Created epg policy {%+v}", gp)
+
+	return gp, nil
+}
+
+// restoreEpgPolicies restores all EPG policies from state store
+func restoreEpgPolicies(stateDriver core.StateDriver) error {
+	// read all epg policies
+	gp := new(EpgPolicy)
+	gp.StateDriver = stateDriver
+	gpCfgs, err := gp.ReadAll()
+	if err == nil {
+		for _, gpCfg := range gpCfgs {
+			epgp := gpCfg.(*EpgPolicy)
+			log.Infof("Restoring EpgPolicy: %+v", epgp)
+
+			// save it in cache
+			epgPolicyDb[epgp.EpgPolicyKey] = epgp
+
+			// Restore all rules within the policy
+			for ruleKey, ruleMap := range epgp.RuleMaps {
+				log.Infof("Restoring Rule %s, Rule: %+v", ruleKey, ruleMap.Rule)
+
+				// delete the entry from the map so that we can add it back
+				delete(epgp.RuleMaps, ruleKey)
+
+				// Add the rule to epg Policy
+				err := epgp.AddRule(ruleMap.Rule)
+				if err != nil {
+					log.Errorf("Error restoring rule %s. Err: %v", ruleKey, err)
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// FindEpgPolicy finds an epg policy
+func FindEpgPolicy(epgpKey string) *EpgPolicy {
+	return epgPolicyDb[epgpKey]
+}
+
+// Delete deletes the epg policy
+func (gp *EpgPolicy) Delete() error {
+	// delete from the DB
+	delete(epgPolicyDb, gp.EpgPolicyKey)
+
+	return gp.Clear()
+}
+
+// createOfnetRule creates a directional ofnet rule
+func (gp *EpgPolicy) createOfnetRule(rule *contivModel.Rule, dir string) (*ofnet.OfnetPolicyRule, error) {
+	ruleID := gp.EpgPolicyKey + ":" + rule.Key + ":" + dir
+
+	// Create an ofnet rule
+	ofnetRule := new(ofnet.OfnetPolicyRule)
+	ofnetRule.RuleId = ruleID
+	ofnetRule.Priority = rule.Priority
+	ofnetRule.Action = rule.Action
+
+	remoteEpgID := 0
+	// See if user specified an endpoint Group in the rule
+	if rule.EndpointGroup != "" {
+		epgKey := rule.TenantName + ":" + rule.Network + ":" + rule.EndpointGroup
+
+		// find the endpoint group
+		epg := contivModel.FindEndpointGroup(epgKey)
+		if epg == nil {
+			log.Errorf("Error finding endpoint group %s", epgKey)
+			return nil, core.Errorf("endpoint group not found")
+		}
+
+		remoteEpgID = epg.EndpointGroupID
+	}
+
+	// Set protocol
+	switch rule.Protocol {
```

Comment:
I've gone thru multiple iteration of this. Code here is little long to avoid complexity in other places.
I'm sure there are other ways of doing it :)